server/cache.go

```diff
@@ -91,394 +89,144 @@ func (s *storesInfo) getStoreCount() int {
-	// search key -> region id
-	searchRegions *regionTree
-	// store id -> region count
-	storeRegionCount map[uint64]uint64
-
-	leaders *leaders
+	tree      *regionTree
+	regions   map[uint64]*regionInfo
+	leaders   map[uint64]map[uint64]*regionInfo
+	followers map[uint64]map[uint64]*regionInfo
 }
 
 func newRegionsInfo() *regionsInfo {
 	return &regionsInfo{
-		regions:          make(map[uint64]*metapb.Region),
-		searchRegions:    newRegionTree(),
-		storeRegionCount: make(map[uint64]uint64),
-		leaders: &leaders{
-			storeRegions: make(map[uint64]map[uint64]struct{}),
-			regionStores: make(map[uint64]uint64),
-		},
+		tree:      newRegionTree(),
+		regions:   make(map[uint64]*regionInfo),
+		leaders:   make(map[uint64]map[uint64]*regionInfo),
+		followers: make(map[uint64]map[uint64]*regionInfo),
 	}
 }
 
-// getRegion gets the region and leader peer by regionKey.
-func (r *regionsInfo) getRegion(regionKey []byte) (*metapb.Region, *metapb.Peer) {
-	r.RLock()
-	defer r.RUnlock()
-
-	region := r.searchRegions.search(regionKey)
-	if region == nil {
-		return nil, nil
-	}
-
-	regionID := region.GetId()
-	leaderStoreID, ok := r.leaders.regionStores[regionID]
-	if ok {
-		return cloneRegion(region), leaderPeer(region, leaderStoreID)
-	}
-
-	return cloneRegion(region), nil
-}
-
-// getRegionByID gets the region and leader peer by regionID.
-func (r *regionsInfo) getRegionByID(regionID uint64) (*metapb.Region, *metapb.Peer) {
-	r.RLock()
-	defer r.RUnlock()
-
+func (r *regionsInfo) getRegion(regionID uint64) *regionInfo {
 	region, ok := r.regions[regionID]
 	if !ok {
-		return nil, nil
-	}
-
-	leaderStoreID, ok := r.leaders.regionStores[regionID]
-	if ok {
-		return cloneRegion(region), leaderPeer(region, leaderStoreID)
+		return nil
 	}
-
-	return cloneRegion(region), nil
+	return region.clone()
 }
 
-// getRegions gets all the regions, returns nil if not found.
-func (r *regionsInfo) getRegions() []*metapb.Region {
-	r.RLock()
-	defer r.RUnlock()
-
-	regions := make([]*metapb.Region, 0, len(r.regions))
-	for _, region := range r.regions {
-		regions = append(regions, cloneRegion(region))
+func (r *regionsInfo) setRegion(region *regionInfo) {
+	if origin, ok := r.regions[region.GetId()]; ok {
+		r.removeRegion(origin)
 	}
-
-	return regions
+	r.addRegion(region)
 }
 
-func (r *regionsInfo) addRegion(region *metapb.Region) {
-	r.searchRegions.insert(region)
-
-	_, ok := r.regions[region.GetId()]
-	if ok {
-		log.Fatalf("addRegion for already existed region in regions - %v", region)
-	}
-
+func (r *regionsInfo) addRegion(region *regionInfo) {
+	// Add to tree and regions.
+	r.tree.update(region.Region)
 	r.regions[region.GetId()] = region
 
-	r.addRegionCount(region)
-}
-
-func (r *regionsInfo) updateRegion(region *metapb.Region) {
-	r.searchRegions.update(region)
-
-	oldRegion, ok := r.regions[region.GetId()]
-	if !ok {
-		log.Fatalf("updateRegion for none existed region in regions - %v", region)
+	if region.Leader == nil {
```

Comment:
If the region exists, we will first remove the original region and add a new one, see `setRegion`.
`addRegion` and `removeRegion` are actually helper functions, only `setRegion` will be used outside.